gateway/blocks_backend.go

```diff
@@ -0,0 +1,779 @@
+	f, err := ufile.NewUnixfsFile(ctx, bb.dagService, nd)
+	// Note: there is an assumption here that non-UnixFS dag-pb should not be returned which is currently valid
+	if err != nil {
+		return md, nil, err
+	}
+
+	if d, ok := f.(files.Directory); ok {
+		dir, err := uio.NewDirectoryFromNode(bb.dagService, nd)
+		if err != nil {
+			return md, nil, err
+		}
+		sz, err := d.Size()
+		if err != nil {
+			return ContentPathMetadata{}, nil, fmt.Errorf("could not get cumulative directory DAG size: %w", err)
+		}
+		if sz < 0 {
+			return ContentPathMetadata{}, nil, fmt.Errorf("directory cumulative DAG size cannot be negative")
+		}
+		return md, NewGetResponseFromDirectoryListing(uint64(sz), dir.EnumLinksAsync(ctx)), nil
+	}
+	if file, ok := f.(files.File); ok {
+		return md, NewGetResponseFromFile(file), nil
+	}
+
+	return ContentPathMetadata{}, nil, fmt.Errorf("data was not a valid file or directory: %w", ErrInternalServerError) // TODO: should there be a gateway invalid content type to abstract over the various IPLD error types?
+}
+
+func (bb *BlocksBackend) GetAll(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	md, nd, err := bb.getNode(ctx, path)
+	if err != nil {
+		return md, nil, err
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	n, err := ufile.NewUnixfsFile(ctx, bb.dagService, nd)
+	if err != nil {
+		return md, nil, err
+	}
+	return md, n, nil
+}
+
+func (bb *BlocksBackend) GetBlock(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.File, error) {
+	md, nd, err := bb.getNode(ctx, path)
+	if err != nil {
+		return md, nil, err
+	}
+
+	return md, files.NewBytesFile(nd.RawData()), nil
+}
+
+func (bb *BlocksBackend) Head(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	md, nd, err := bb.getNode(ctx, path)
+	if err != nil {
+		return md, nil, err
+	}
+
+	rootCodec := nd.Cid().Prefix().GetCodec()
+	if rootCodec != uint64(mc.DagPb) {
+		return md, files.NewBytesFile(nd.RawData()), nil
+	}
+
+	// TODO: We're not handling non-UnixFS dag-pb. There's a bit of a discrepancy
+	// between what we want from a HEAD request and a Resolve request here and we're using this for both
+	fileNode, err := ufile.NewUnixfsFile(ctx, bb.dagService, nd)
+	if err != nil {
+		return ContentPathMetadata{}, nil, err
+	}
+
+	return md, fileNode, nil
+}
+
+func (bb *BlocksBackend) GetCAR(ctx context.Context, p ImmutablePath, params CarParams) (io.ReadCloser, error) {
+	contentPathStr := p.String()
+	if !strings.HasPrefix(contentPathStr, "/ipfs/") {
+		return nil, fmt.Errorf("path does not have /ipfs/ prefix")
+	}
+	firstSegment, _, _ := strings.Cut(contentPathStr[6:], "/")
+	rootCid, err := cid.Decode(firstSegment)
+	if err != nil {
+		return nil, err
+	}
+
+	r, w := io.Pipe()
+	go func() {
+		cw, err := storage.NewWritable(w, []cid.Cid{rootCid}, car.WriteAsCarV1(true))
```

Comment:
TODO: either replace `rootCid` with the CID of final path segment or pass empty list (see https://github.com/ipfs/specs/pull/402#discussion_r1221517832).
The former is preferable, as it will keep the old behavior (before IPIP-402).